File: src/main/java/gameplay/physics/GameMapHelper.java
package gameplay.physics;

import gameplay.scene.Floor;
import gameplay.scene.GameMap;
import lombok.RequiredArgsConstructor;
import util.Validate;

@RequiredArgsConstructor
public class GameMapHelper {
    private final GameMap map;
    private final gameplay.Character character;

    enum FloorIs {Below, Above}

    enum From {Left, Right}

    double closestDistance(FloorIs floorIs) {
        double minDistance = map.getHeight();  // largest possible value

        for (Floor floor : map.getFloors()) {
            if ((floor.getLeft() <= character.getRightSideX()) && (character.getLeftSideX() <= floor.getRight())) {
                double off = (floorIs == FloorIs.Below)
                        ? character.getY() - floor.getTop()
                        : floor.getBottom() - character.getHeadY();

                if (off == 0.0) {
                    return 0;
                }
                if (off > 0.0) {
                    if (off < minDistance) {
                        minDistance = off;
                    }
                }
            }
        }
        return minDistance;
    }

    double closestHorizontalObstacle(From from) {

        double minDistance = map.getWidth();  // largest possible value

        for (Floor floor : map.getFloors()) {
            if (Validate.inside(floor.getBottom(), character.getFeetY() - Floor.HEIGHT, character.getHeadY())) {
                double dist = (from == From.Right)
                        ? floor.getLeft() - character.getRightSideX()
                        : character.getLeftSideX() - floor.getRight();

                if (dist == 0) {
                    return 0;
                }
                if (dist > 0) { // only change distance if Floor is in the right direction (negative values are opposite direction)
                    minDistance = Math.min(minDistance, dist);
                }
            }
        }
        return minDistance;
    }
}

File: src/main/kotlin/gameplay/physics/GameMapHelper.kt
package gameplay.physics

import gameplay.Character
import gameplay.scene.Floor
import gameplay.scene.GameMap

class GameMapHelper(val map: GameMap, val character: Character) {
    fun closest(from: From): Double {
        return map.floors
                .stream()
                .filter { from.overlaps(character, it) }
                .mapToDouble { from.distance(character, it) }
                .filter { it >= 0.0 }
                .min()
                .orElseGet { from.max(map).toDouble() }
    }

    enum class From(val overlaps: (Character, Floor) -> Boolean, val max: (GameMap) -> Int, val distance: (Character, Floor) -> Double) {

        Left(Overlaps::horizontally, GameMap::width, { character, floor -> character.leftSideX - floor.right }),
        Right(Overlaps::horizontally, GameMap::width, { character, floor -> -(character.rightSideX - floor.left) }),
        Below(Overlaps::vertically, GameMap::height, { character, floor -> character.y - floor.top }),
        Above(Overlaps::vertically, GameMap::height, { character, floor -> -(character.headY - floor.bottom) });
    }

    companion object Overlaps {
        fun vertically(character: Character, floor: Floor): Boolean {
            return character.rightSideX >= floor.left && floor.right >= character.leftSideX
        }

        fun horizontally(character: Character, floor: Floor): Boolean {
            return character.headY > floor.bottom && floor.top > character.feetY
        }
    }
}
